File: app/src/main/java/ru/alexgiltd/musicdb/presentation/artist/ArtistListPresenter.java
package ru.alexgiltd.musicdb.presentation.artist;

import android.util.Log;

import com.arellomobile.mvp.InjectViewState;

import java.util.ArrayList;
import java.util.List;

import javax.inject.Inject;

import io.reactivex.Observable;
import io.reactivex.android.schedulers.AndroidSchedulers;
import io.reactivex.disposables.Disposable;
import io.reactivex.schedulers.Schedulers;
import ru.alexgiltd.musicdb.data.repository.Repository;
import ru.alexgiltd.musicdb.model.ArtistModel;
import ru.alexgiltd.musicdb.util.BasePresenter;

@InjectViewState
public class ArtistListPresenter extends BasePresenter<ArtistListView> {

    private static final String TAG = ArtistListPresenter.class.getSimpleName();

    private final Repository repository;

    private final List<ArtistModel> artists = new ArrayList<>();

    @Inject
    public ArtistListPresenter(Repository repository) {
        this.repository = repository;
    }

    @Override
    protected void onFirstViewAttach() {

        loadArtistList();
    }

    public void onItemClicked(ArtistModel artistDetailsModel) {
        getViewState().openArtistDetails(artistDetailsModel);
    }

    private void loadArtistList() {

        Disposable disposable = repository.getArtists(10)
                .subscribeOn(Schedulers.io())
                .flatMap(simpleArtistModels -> Observable.fromIterable(simpleArtistModels)
                        .concatMap(artist -> artist.getMbid().isEmpty()
                                ? repository.getArtistDetailsByName(artist.getName()).toObservable().subscribeOn(Schedulers.io())
                                : repository.getArtistDetailsByMbid(artist.getMbid()).toObservable().subscribeOn(Schedulers.io())
                        )
                )
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe(
                        artistModel -> {
                            artists.add(artistModel);
                            getViewState().showArtists(new ArrayList<>(artists));
                        },
                        error -> {
                            getViewState().onFinishLoading();
                            getViewState().showError(error.getMessage());
                            Log.e(TAG, "loadArtistList(): ", error);
                        },
                        () -> {
                            getViewState().onFinishLoading();
                            Log.d(TAG, "addArtists(): onComplete()");
                        },
                        subscription -> {
                            getViewState().onStartLoading();
                        }
                );

        unsubscribeOnDestroy(disposable);

    }

}

File: app/src/main/java/ru/alexgiltd/musicdb/presentation/artist/ArtistListPresenter.kt
package ru.alexgiltd.musicdb.presentation.artist

import com.arellomobile.mvp.InjectViewState
import io.reactivex.Observable
import io.reactivex.android.schedulers.AndroidSchedulers
import io.reactivex.schedulers.Schedulers
import ru.alexgiltd.musicdb.data.repository.Repository
import ru.alexgiltd.musicdb.model.ArtistModel
import ru.alexgiltd.musicdb.util.BasePresenter
import timber.log.Timber
import java.util.*
import javax.inject.Inject

@InjectViewState
class ArtistListPresenter @Inject constructor(private val repository: Repository)
    : BasePresenter<ArtistListView>() {

    private val artists: MutableList<ArtistModel> = ArrayList()

    override fun onFirstViewAttach() {
        loadArtists()
    }

    fun onItemClicked(artistDetailsModel: ArtistModel) {
        viewState.openArtistDetails(artistDetailsModel)
    }

    private fun loadArtists() {
        val disposable = repository.getArtists(10)
                .flatMap { list ->
                    Observable.fromIterable(list)
                            .concatMap { simpleArtistModel ->
                                if (simpleArtistModel.mbid.isEmpty())
                                    repository.getArtistDetailsByName(simpleArtistModel.name)
                                            .subscribeOn(Schedulers.io())
                                else
                                    repository.getArtistDetailsByMbid(simpleArtistModel.mbid)
                                            .subscribeOn(Schedulers.io())
                            }
                }
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .doOnSubscribe { viewState.onStartLoading() }
                .subscribe(this::doOnNext, this::doOnError, this::doOnComplete)

        unsubscribeOnDestroy(disposable)
    }

    private fun doOnNext(artistModel: ArtistModel) {
        artists.add(artistModel)
        viewState.showArtists(ArrayList(artists))
    }

    private fun doOnError(throwable: Throwable) {
        viewState.onFinishLoading()
        viewState.showError(throwable.message!!)
        Timber.e(throwable, "loadArtists(): ")
    }

    private fun doOnComplete() {
        viewState.onFinishLoading()
    }
}
